Utils/API/server/lib/dkb/api/storages/es/query/deriv

```diff
@@ -0,0 +1,54 @@
+{
+  "size": 0,
+  "query": {
+    "bool": {
+      "must": [
+        {"terms": {"primary_input": ["aod","rpvll"]}},
+        {"term": {"project": %(project)s}},
+        {"terms": {"ctag": %(ctag)s}},
+        {"term": {"status": "done"}},
+        {"range": {"input_bytes": {"gt": 0}}},
+        {"has_child": {
+            "type": "output_dataset",
+            "query" : {
+              "bool": {
+                "must": [
+                  {"term": {"data_format": %(output)s}},
+                  {"range": {"bytes": {"gt": 0}}}
```

Comment:
Seems so. For example, task with id `14120127` has a single deleted dataset `data18_13TeV.00349646.physics_Main.deriv.DAOD_BPHY7.f933_m1960_p3553_tid14120127_00` - it is not found by the query, but IS found by one with this restriction removed.

Reply:
Sorry, what I meant to ask was: is it the expected behaviour?
Those input datasets, filtered by this condition, are excluded from the resulting `"input_{events|bytes}"` -- did you think about it and decided that it its correct (then I'm asking about motivation for this decision), or was this effect unintended (then I'm asking you to take a closer look at it and make a deliberate decision)?

Reply:
Hm. This condition was, like the rest of the query, taken from ProdTask's code (so the answer to your question is "no, I was not specifically evaluating this condition"). Should we change it?

While the input datasets are, as you correctly say, excluded from resulting `input` values, the output datasets of the corresponding tasks are similarly excluded from the `result` ones. Therefore, there seems to be no error in this regard.

Another point can be "why should both input and result values of these tasks be excluded from the statistics?", but, if I understand correctly, the answer is "because without input/output/both datasets no values - or only one of each pair - can be obtained, making the resulting calculation (for this particular task) impossible".

Reply:
> This condition was, like the rest of the query, taken from ProdTask's code

My point here is that even if you have someone else's code to get the general idea from, it is usually a good idea to dive into it and understand what the copied pieces of code do and why/for what reason. Right, Mikhail knows better what he needs to see in the result -- but everyone can make a mistake, so if you have no idea why something is done this way and not another, it may as well mean that it *is* a mistake and here's a chance to fix it. Or not, but in order to find out you have to ask the question.

> (so the answer to your question is "no, I was not specifically evaluating this condition"). Should we change it?

And that's the thing that should definitely be changed -- the belief that others can do no mistake ;)

---

Now, to the question itself...

> While the input datasets are, as you correctly say, excluded from resulting input values, the output datasets of the corresponding tasks are similarly excluded from the result ones. Therefore, there seems to be no error in this regard.

Sounds reasonable for me: if you have excluded some task from the calculation, it equally affects `sum(input)` and `sum(oputput)`, so that the result ratio should be correct.

> "why should both input and result values of these tasks be excluded from the statistics?" -- "because without input/output/both datasets no values - or only one of each pair - can be obtained, making the resulting calculation (for this particular task) impossible"

But we do not perform "calculation for particular task" -- it's two total sums and then the ratio of these values. Yet if there was only one task matching the `project-tag-format` combination -- that'd be a problem indeed...

I'd say that the original idea might have been as follows: if we know values only for output and add them to the total sum, the `total(output)` would contain more tasks than `total(input)` and dividing one by another would simply be incorrect (and vice versa). Does it sound logical, or is a bit of a stretch?

Reply:
Aha, it seems that I've missed your point. While there are "incorrect" tasks with 0 input and/or output bytes and these tasks are discarded because their contribution cannot be calculated correctly, there is (by logic) a possibility of a "correct" task with positive input but zero output [1]. To give an abstract example, if input is the list `[1, 2, 3]`, then "incorrect" task A with algorithm "find all even values" (or any other one) can have output `[]` instead of expected `[2]` due to some error/output being lost/whatever, but a "correct" task B with algorithm "find all negative values" can produce the same output `[]`, and there is nothing wrong with it. In such case, input events/bytes of B should be added to input total, which currently is not done. Is this what you've been trying to tell me?

If the answer is "yes", then the next question is "how can we distinguish between tasks A and B?". I don't have an answer, so we should probably discuss this with Mikhail, or someone else who is better informed about the ATLAS' data.

Some comments on other points of interest:

> > This condition was, like the rest of the query, taken from ProdTask's code
> 
> My point here is that even if you have someone else's code to get the general idea from, it is usually a good idea to dive into it and understand what the copied pieces of code do and why/for what reason. Right, Mikhail knows better what he needs to see in the result -- but everyone can make a mistake, so if you have no idea why something is done this way and not another, it may as well mean that it _is_ a mistake and here's a chance to fix it. Or not, but in order to find out you have to ask the question.
> 
> > (so the answer to your question is "no, I was not specifically evaluating this condition"). Should we change it?
> 
> And that's the thing that should definitely be changed -- the belief that others can do no mistake ;)

Sure, I agree with these two points. My "should we change it?" was meant as "So, you have an idea of some sort that suggests that we should change it? Let's hear it then." (if my guess above is correct, then you've already voiced the idea, but I missed it), not as "Why should we bother changing it if it was already decided by someone else?".

> > "why should both input and result values of these tasks be excluded from the statistics?" -- "because without input/output/both datasets no values - or only one of each pair - can be obtained, making the resulting calculation (for this particular task) impossible"
> 
> But we do not perform "calculation for particular task" -- it's two total sums and then the ratio of these values.

True, I've misworded my statement.

> Yet if there was only one task matching the `project-tag-format` combination -- that'd be a problem indeed...

Cases of having zero input/no tasks at all/etc are taken care of in the code, with `try/except` and `if`s. This can be improved, because the current version simply assigns 0 to some/all values, but at least in some form this protective mechanism does exist.

> I'd say that the original idea might have been as follows: if we know values only for output and add them to the total sum, the `total(output)` would contain more tasks than `total(input)` and dividing one by another would simply be incorrect (and vice versa). Does it sound logical, or is a bit of a stretch?

This is more or less what I've wanted to say above, with "particular task". Except that my logic was "we can't add unknown value to one of the two sums, which is wrong" instead of "if we add value only to one of the two sums, we upset the input-output equality, which is wrong", which is more or less the same. This returns us to the beginning, problem of distinguishing between two types of tasks and their values, a.k.a. how to separate "zero which is zero" from "zero which is unknown value".

[1] By abstract logic, it is also possible to have a task producing output from "nothing", with zero input but positive output. This case, however, shouldn't exist in here, in ATLAS data.

Reply:
Fine, now I guess we get to the bottom of my original question.

---

>next question is "how can we distinguish between tasks A and B?"

where
>"incorrect" task A <...> can have output `[]` instead of expected `[2]` due to some error <...>, but a "correct" task B <...> can produce the same output `[]`, and there is nothing wrong with it.

My guess (and nothing but a guess):
* if dataset exists -- its files contain some metadata, so size on disks will not be zero;
* if task derives "empty set of data" from the input, it may not produce any dataset at all.

For now I suggest to leave this one aside to discuss it with Mikhail on occasion (at the meeting or by e-mail).

If you're OK with this, please add a note/card in Trello (to not forget about your question) and resolve this conversation.

Reply:
https://trello.com/c/bMcR3PkQ